### PR TITLE
[PhpUnitBridge] install PHPUnit 6 on PHP 7.2

### PR DESCRIPTION
**Pull request by xabbuh**
| Q             | A
| ------------- | ---
| Branch?       | 3.4
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #23943
| License       | MIT
| Doc PR        |

### CODE REVIEW

**Comment by dmaicher** [comment]
phpunit 6 itself is not really backwards compatible due to the namespace changes. Does the phpunit bridge add any backwards compat layer for it? Or will my tests just fail if my testsuite is running on phpunit 5 without any namespaced classes?

File: src/Symfony/Bridge/PhpUnit/bin/simple-phpunit [hunk 1]
```diff
@@ -15,8 +15,17 @@
...
-$PHPUNIT_VERSION = PHP_VERSION_ID >= 50600 ? getenv('SYMFONY_PHPUNIT_VERSION') ?: '5.7' : '4.8';
+if (PHP_VERSION_ID >= 70200) {
+    // PHPUnit 6 is required for PHP 7.2+
+    $PHPUNIT_VERSION = getenv('SYMFONY_PHPUNIT_VERSION') ?: '6.3';
+} elseif (PHP_VERSION_ID >= 50600) {
+    // PHPUnit 4 does not support PHP 7
+    $PHPUNIT_VERSION = getenv('SYMFONY_PHPUNIT_VERSION') ?: '5.7';
+} else {
+    // PHPUnit 5.1 requires PHP 5.6+
+    $PHPUNIT_VERSION = getenv('SYMFONY_PHPUNIT_VERSION') ?: '4.8';
```
Comment (stof) [inline src/Symfony/Bridge/PhpUnit/bin/simple-phpunit]: This changes the existing behavior. For PHP < 5.6, the env variable was ignored before, forcing to use 4.8, as this is the only usable version there:

- newer versions are not supported on these PHP runtimes
- older versions than 4.8 are not supported by the bridge.

I would keep this behavior (anyway, changing it is a BC break).

**Review by stof — CHANGES_REQUESTED** [review]

**Comment by stof** [comment]
@dmaicher if you run your testsuite on PHPUnit 6 while using `PHPUnit_Framework_TestCase` as base class, things will break totally.
However, using the PHPUnit bridge with PHPUnit 5.7 on PHP 7.2 will also make your CI red: the bridge makes the run fail when encountering deprecations, and PHPUnit 5.7 triggers a deprecation coming from PHP 7.2.

However, making your testsuite compatible with several PHPUnit versions is possible. PHPUnit 4.8.35 and 5.4.3+ are providing the `PHPUnit\Framework\TestCase` class as a forward compatibility layer, allowing you to migrate to the namespace API in your `extends` clause. This is the way to go.
And note that PHPUnit 6 is the only actively maintained version of PHPUnit, so migrating to it is a must-have IMO.
And the migration to namespaced base testcase can be done easily. If you don't care about getting `extends \PHPUnit\Framework\TestCase` instead of using a use statement, it is a simple search&replace.

File: src/Symfony/Bridge/PhpUnit/bin/simple-phpunit (hunk 1 above)
Comment (xabbuh) [inline src/Symfony/Bridge/PhpUnit/bin/simple-phpunit]: You are right. I misread the nested ternary operators.

**Review by nicolas-grekas — APPROVED** [review]

**Comment by Simperfit** [comment]
Maybe it will help for #23671, 👍

**Review by Simperfit — APPROVED** [review]

**Review by Tobion — APPROVED** [review]

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
I think this should be merged in 3.3 instead.

**Comment by nicolas-grekas** [comment]
:+1: for 3.3

**Comment by xabbuh** [comment]
rebased on 3.3

**Comment by fabpot** [comment]
Thank you @xabbuh.

File: src/Symfony/Bridge/PhpUnit/CHANGELOG.md [hunk 2]
```diff
@@ -1,6 +1,11 @@
 CHANGELOG
 =========
 
+3.4.0
```
Comment (fabpot) [inline src/Symfony/Bridge/PhpUnit/CHANGELOG.md]: removed after the merge (2230e31727)

Comment (xabbuh) [reply src/Symfony/Bridge/PhpUnit/CHANGELOG.md]: thanks

**Pull request merged**